### ISSUE TITLE
Universal Connect-OAuthTwitter Suggestion

### ISSUE THREAD

**Issue by mkellerman** [CollinChaffin/PSTwitter#4]
@CollinChaffin 

Instead of having to build a switch and multiple parameters for the OAuthTwitter, couldn't we simply do something like this:

$oAuthRequestString = (Connect-OAuthTwitter $psTwitterEndpointTweet "status=hello world");

Sometype of universal authorization string request?

**Comment by CollinChaffin** [comment]
I wrote it to be modular and use PSObject and with DM, "status" isn't the only thing needed so you can't make it universal.  What is $psTwitterEndpointTweet anyway?  If I need to add a couple more "types" to query I don't think that adds that much complexity and still keeps it logical enough that you can follow the code.

Feel free to shoot an example but the examples I saw before I rewrote tried to do exactly that and the one routine "connect-oauthtwitter" became kind of a beast where if Twitter changed something, this massive single routine needed updating so I tore it all down and rewrote it this way.  This way, signature is it's own modular routine, noonce it's own, etc.

What is it exactly you are trying to add, the query pieces?  I will begin looking at that soon as well and shoot out an example of how I might do it and we can compare methods to see which makes more sense.

Thanks!

**Comment by mkellerman** [comment]
If you have 10 minutes, i could post something up… 

I’ve created a different version of your function.. just to see what it could look like..

**Comment by mkellerman** [comment]
It’s up..

Please look at: 

Connect-OAuthTwitterUniversal 
New-TwitterOAuthSignature2
Get-TwitterFollowers
Get-TwitterFriends 

So sorry for the mess.. i started this 1h ago.. havent had time to clean it up much..

**Comment by CollinChaffin** [comment]
You should really do a  branch/pull request then the difff can show what was changed to avoid having to read and compare yourself.....kinda point of github :)  So are these new functions in your repo?

**Comment by mkellerman** [comment]
Yeah, i’m sorry.. i dont know what i’m doing.. :P

So, each time, i branch/pull request?

**Comment by CollinChaffin** [comment]
No, you branch once then when you commit I approve the push to code base.  My concern with what you have above is that remember connect-oauth function in case of DM requires diff info, message (not called status) AND who it's TO.  This is required to then inside that func be passed to signature (which requires for crypto to have those pieces) hence why I was telling connect-oauth what it's operation was when calling from tweet, DM, (new calls) and param set because those sets of info are required as it calls to generate signature and string.  Make sense?

**Comment by mkellerman** [comment]
Kinda.. 

you could do:

$psTwitterParameters = "screen_name=mkellerman&text=helloworld"

$oAuthRequestString = (Connect-OAuthTwitterUniversal $psTwitterEndpointDirectMessage $psTwitterParameters )

This would still work, and wouldn't need the Switch you use in Connect-OAuth and Signature.. 

Especially if you want to keep the same function with GET & POST methods.. 

What do you think?

**Comment by CollinChaffin** [comment]
If you note inside the New-TwitterOAuthSignature called from within the Connect-OAuth:

'Tweeting'
                {
                    $oAuthSignatureBase += [System.Uri]::EscapeDataString("status=$($objOAuth.PSObject.Properties["oauth_tweetmessage"].Value)")
                }
                'Direct'
                {
                    $oAuthSignatureBase += [System.Uri]::EscapeDataString("screen_name=$($objOAuth.PSObject.Properties["oauth_directrecipient"].Value)&")
                    $oAuthSignatureBase += [System.Uri]::EscapeDataString("text=$($objOAuth.PSObject.Properties["oauth_directmessage"].Value)")  
                }

So, on a "universal" connect-oauth one way or another you still need to detect which operation you are performing (tweet, dm, etc) be it param set or another way.  I like populating one single PSObject with all info for the operation and calling the connect from tweet, dm, new queries for followers, and simply passing one param indicating which func and required info.  Then, for future growth, inside connect-oauth you simply add one new param set with those new info pieces, add those info to the already set up PSObject, and since the sig and others get passed the entire object, only minor code changes/additions would ever be required to say add a signature generation for a query - again add the one new param set in that sub-function and cooresponding code and off you go.

I've had a week from hell so I'll take you're code and study it and leave this issue open and then post what I would propose and we can discuss what makes more sense!

**Comment by mkellerman** [comment]
Did you look at my New-TwitterOAuthSignature2 ?

By sending that entire $psTwitterParameters, it still works.. 

$oAuthSignatureBase += [System.Uri]::EscapeDataString("$($objOAuth.PSObject.Properties["oauth_parameters"].Value)")

**Comment by mkellerman** [comment]
Regarding your week from hell.. terribly sorry.. i didn't want to cause havoc with your module.. :P 

Take your time.. i'll clean it up more.. and make it easier to read.

I'll do a fresh pull/request.. and modify your code directly instead of create new functions.. so you can compare it to your code directly..

In my mind.. its easier to handle all the code changes inside of Get-TwitterFollower to handle what it's supposed to do.. (by changing the $psTwitterParameters).. then having to had parameters in Connect-OAuthTwitter & New-TwitterOAuthSignature each time..

**Comment by CollinChaffin** [comment]
No problem and don't mean to come across difficult I just looked at it and now I get what you're getting at regarding POST vs GET.  The only thing I'd say is I see you tested your function with new follower/friends but looks like at a glance original Tweet and Direct still using old so maybe ensure those original functions will be okay and what they would need to look like to send a tweet or DM using new function.  Hopefully tomorrow I will have a little more time to look at it!  Thanks!

**Comment by mkellerman** [comment]
Ill fix the DM and the tweet.. And make sure it all works with the new functions..  See if that simplifies or complicates thing in the event Twitter changes their API calls

**Comment by mkellerman** [comment]
I ‘think’ i created a new branch called Universal-Functions..

And posted it there.. all 4 functions (DM, Tweet, Followers, Friends) work now.. and would be very simple to add new functions without touching OAuth and Signature…

Tell me what you think.

PS: I’m having issues when sending multiple parameters (screen_name & text).. sometimes it works, sometimes gives me Authorization 401 error..  This is the same error when i add that second parameter for the Followers/Friends (Cursor position).. i’m not quite familiar with the Twitter API, so maybe something you could take a peak as well.. 

Anyways, thanks for your help on this! I love your work.. =)

**Comment by mkellerman** [comment]
Updated a new version.. let me know if you see it.. version 127.0.0.1 (yeah, i have no idea what i'm doing or how i should be calling them).. lol

**Comment by mkellerman** [comment]
So.. i couldn't sleep... and finally got everything working.. 

bugs resolved:
- sorting the parameters
- having multiple parameters
- cursor paging
